src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java

```diff
@@ -0,0 +1,530 @@
+    }
+
+    /**
+     * Struct to maintain the state of the current file being streamed in.
+     */
+    static class FileTransferContext
+    {
+        /**
+         * A queue for the incoming {@link ByteBuf}s, which will be processed by the {@link #blockingIOThread}.
+         *
+         * // TODO:JEB confirm this next comment
+         * Items in this queue
+         * live longer than {@link StreamingInboundHandler#pendingBuffers}, and hence we set the low/high water marks from this instance.
+         */
+        final AppendingByteBufInputPlus inputStream;
+        private final Type type;
+
+        int headerLength;
+        FileMessageHeader header;
+        StreamSession session;
+        long remaingPayloadBytesToReceive;
+        ColumnFamilyStore cfs;
+
+        /**
+         * If the target file is using sstable compression, this is the index into the header's {@link CompressionInfo#chunks}
+         * that is currently being operated on.
+         */
+        private int currentCompressionChunk;
+
+        private FileTransferContext(ChannelConfig config, Type type)
+        {
+            this.type = type;
+            if (type.isTransfer())
+                inputStream = new AppendingByteBufInputPlus(AUTO_READ_LOW_WATER_MARK, AUTO_READ_HIGH_WATER_MARK, config);
+            else
+                inputStream = null;
+        }
+    }
+
+    /**
+     * A task that can execute the blocking deserialization behavior of {@link StreamReader#read(ColumnFamilyStore, AppendingByteBufInputPlus)}.
+     */
+    private class DeserializingSstableTask implements Runnable
+    {
+        private final ChannelHandlerContext ctx;
+        private final InetSocketAddress remoteAddress;
+        private final BlockingQueue<FileTransferContext> queue;
+
+        DeserializingSstableTask(ChannelHandlerContext ctx, InetSocketAddress remoteAddress, BlockingQueue<FileTransferContext> queue)
+        {
+            this.ctx = ctx;
+            this.remoteAddress = remoteAddress;
+            this.queue = queue;
+        }
+
+        @Override
+        public void run()
+        {
+            FileMessageHeader header;
+            try
+            {
+                while (state != State.CLOSED)
+                {
+                    FileTransferContext transferContext = queue.poll(1, TimeUnit.SECONDS);
+                    if (transferContext == null)
+                        continue;
+
+                    header = transferContext.header;
+
+                    StreamReader reader = new StreamReader(header, transferContext.session);
+                    SSTableMultiWriter ssTableMultiWriter = reader.read(transferContext.cfs, transferContext.inputStream);
+                    transferContext.session.receive(header, ssTableMultiWriter);
+                }
+            }
+            catch (InterruptedException e)
+            {
+                // nop, thread was interrupted by the parent class (this is, or should be, normal/good/happy path)
```

Comment:
The way around that is can be to close the channel or file or whatever rather than interrupt. If you interrupt in Java I think it actually closes the FD in at least some cases (yuck!).